Blazor Overview updates (#12106)
aspnet/AspNetCore.Docs: aspnetcore/blazor/index.md

description: Explore ASP.NET Core Blazor, a way to build interactive client-side web UI with .NET in an ASP.NET Core app.
monikerRange: '>= aspnetcore-3.0'
ms.author: riande
ms.custom: "mvc, seoapril2019"
ms.date: 04/22/2019
uid: blazor/index
---
# Introduction to Blazor

By [Daniel Roth](https://github.com/danroth27) and [Luke Latham](https://github.com/guardrex)

*Welcome to Blazor!*

> Build interactive client-side web UI with .NET:

Reviewer: ```suggestion
Blazor is a framework for building interactive client-side web UI with .NET. 

With Blazor you can:
```

* Create rich interactive UIs using C# instead of JavaScript.
* Share server-side and client-side app logic written with .NET.
* Render the UI as HTML and CSS for wide browser support, including mobile browsers.

> Blazor supports app development scenarios required by most apps:

Reviewer: This Blazor feature list feels a bit awkward here. Some of these features are self-explanatory, but many of them are not.

I think it's important that this first section covers: 1. What is Blazor? and 2. Why should I care?

I think we have (1) covered, but I think we could do a better job with (2). We have some text that talks about the advantage of using Blazor in the client-side section below. How about we move it up here in place of this feature list?

* Event handling
* Data binding
* Routing
* Dependency injection
* Layouts
* Templates
* Cascading values
* Parameters

## Components

> A *component* in Blazor is an element of UI, such as a page, dialog, or data entry form. Components handle user events and define flexible UI rendering logic. Components can be nested and reused.

Reviewer: ```suggestion
Blazor apps are based on *components*. A component in Blazor is an element of UI, such as a page, dialog, or data entry form. Components handle user events and define flexible UI rendering logic. Components can be nested and reused.
```

```cshtml
<div>
    <h1>@Title</h1>

    @ChildContent

    <button onclick="@OnYes">Yes!</button>
</div>

@functions {
    [Parameter]
    public string Title { get; set; }

    [Parameter]
    public RenderFragment ChildContent { get; set; }

    private void OnYes()
    {
        Console.WriteLine("Write to the console in C#!");
    }
}
```

The dialog's body content (`ChildContent`) and title (`Title`) are provided by the component that uses this component in its UI. `OnYes` is a C# method triggered by the button's `onclick` event.

Blazor uses natural HTML tags for UI composition. HTML elements specify components, and a tag's attributes pass values to a component's properties. `ChildContent` and `Title` are set by the component that uses the Dialog component (*Index.razor*):

```cshtml
@page "/"

<h1>Hello, world!</h1>

Welcome to your new app.

<Dialog Title="Blazor">
    Do you want to <i>learn more</i> about Blazor?
</Dialog>
```

The dialog is rendered when the parent (*Index.razor*) is accessed in a browser:

![Dialog component rendered in the browser](index/_static/dialog.png)

> When this component is used in the app, [Visual Studio](https://visualstudio.microsoft.com/vs/) IntelliSense speeds development with syntax and parameter completion.

Reviewer: ```suggestion
When this component is used in the app, IntelliSense in [Visual Studio](https://visualstudio.microsoft.com/vs/) and [Visual Studio Code](https://code.visualstudio.com) speeds development with syntax and parameter completion.
```

Components render into an in-memory representation of the browser DOM called a *render tree* that's used to update the UI in a flexible and efficient way.

## Blazor client-side

Blazor client-side is a single-page app framework for building interactive client-side web apps with .NET. Blazor client-side uses open web standards without plugins or code transpilation and works in all modern web browsers, including mobile browsers.

> Using .NET in the browser for client-side web development offers the following advantages:

Reviewer: This is the list I think we should move into the opening section
